src/java.base/share/classes/jdk/internal/vm/CDSParallelPreProcessor.java

```diff
@@ -116,22 +111,99 @@ public static int preLoadAndProcess(String classListFile,
+    static boolean loadClassFromSource(CDSClassListLine info) {
+        if (info.superId == CDSClassListLine.INVALID_ID) {
+            throw new RuntimeException("If source location is specified, super class must be also specified");
+        }
+        if (info.thisId == CDSClassListLine.INVALID_ID) {
+            throw new RuntimeException("If source location is specified, id must be also specified");
+        }
+        if (info.className.startsWith("java.")) {
+            System.err.printf("Prohibited package for non-bootstrap classes: %s from %s\n", info.className, info.source);
+            return false;
+        }
+        preProcessClassWithSource(info.className, info.source, info.interfaces.size());
+        return true;
+    }
+
+    static CDSClassListLine parseClassListLine(String line) {
+        line = line.strip();
+        if (line.startsWith("#")) {
+            // Line starts with '#' is treated as comment and is ignored.
+            return null;
+        } else if (line.startsWith("[")) {
+            // Array is not support in classlist.
+            System.out.println("Preload Warning: Cannot find " + line);
+            return null;
+        } else {
+            // Otherwise, it's a instance klass
+            String[] tokens = line.split(" ");
+            if (tokens.length < 1) {
+                throw new RuntimeException("class name is absent");
+            }
+            try {
+                int tokenCnt = 0;
+                CDSClassListLine info = new CDSClassListLine();
+                info.className = tokens[tokenCnt++];
+                info.className = info.className.replace('/', '.');
+                while (tokenCnt < tokens.length) {
+                    switch (tokens[tokenCnt]) {
+                        case "id:":
+                            info.thisId = Integer.parseInt(tokens[++tokenCnt]);
+                            break;
+                        case "super:":
+                            info.superId = Integer.parseInt(tokens[++tokenCnt]);
+                            break;
+                        case "interfaces:":
+                            tokenCnt++;
+                            while (tokenCnt < tokens.length && isNumeric(tokens[tokenCnt])) {
+                                info.interfaces.add(Integer.parseInt(tokens[tokenCnt]));
+                                tokenCnt++;
+                            }
+                            break;
+                        case "souce:":
+                            info.source = tokens[++tokenCnt];
+                            break;
+                        default:
+                            throw new RuntimeException("Unknown input");
+                    }
+                    tokenCnt++;
+                }
+                return info;
+            } catch (Exception e) {
+                throw new RuntimeException("Invalid format of classlist line: \n\t\"" + line + "\"");
+            }
+        }
+    }
```

Comment:
Could you please use a separate class (i.e. CDSClassListParser.java?) for all the class list format parsing code? Then the details for handling the class list format for custom ClassLoader can be hidden from the parallel processor.